File: introducao-a-computacao/2022/trabalho-pratico/c/application/main.c
#include <stdlib.h>
#include "resultado.h"
#include "contador.h"
#include "io.h"

static int *alocar_numeros(int n)
{
    if (n <= 0)
        return NULL;
    return (int *)malloc((size_t)n * sizeof(int));
}

int main(void)
{
    int quantidade;

    io_print("Quantos números deseja inserir? ");
    fflush(stdout);
    if (!io_ler_int(&quantidade)) {
        io_erro("Erro: entrada inválida. Informe um número inteiro.\n");
        return EXIT_FAILURE;
    }
    if (quantidade <= 0) {
        io_erro("Erro: a quantidade deve ser maior que zero.\n");
        return EXIT_FAILURE;
    }

    int *numeros = alocar_numeros(quantidade);
    if (numeros == NULL) {
        io_erro("Erro: memória insuficiente.\n");
        return EXIT_FAILURE;
    }

    for (int i = 0; i < quantidade; i++) {
        printf("Digite o %dº número: ", i + 1);
        fflush(stdout);
        if (!io_ler_int(&numeros[i])) {
            io_erro("Erro: número inválido na posição.\n");
            free(numeros);
            return EXIT_FAILURE;
        }
    }

    ResultadoContagem resultado = contar_pares_impares(numeros, quantidade);
    free(numeros);

    io_exibir_resultado(&resultado);
    return EXIT_SUCCESS;
}

File: introducao-a-computacao/2022/trabalho-pratico/python/application/main.py
import sys
from pathlib import Path


sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from domain.contador import contar_pares_impares, formatar_resultado
from infrastructure.io import ler_quantidade, ler_numeros, print_erro


def run() -> int:
    q = ler_quantidade()
    if q is None:
        print_erro("Erro: entrada inválida. Informe um número inteiro.")
        return 1
    if q <= 0:
        print_erro("Erro: a quantidade deve ser maior que zero.")
        return 1

    numeros = ler_numeros(q)
    if numeros is None:
        print_erro("Erro: entrada inválida. Informe um número inteiro.")
        return 1

    resultado = contar_pares_impares(numeros)
    print(formatar_resultado(resultado))
    return 0


def main() -> None:
    sys.exit(run())


if __name__ == "__main__":
    main()
